Animation system updated, some more boot files decompiled (+6%), z_fcurve_data_skelanime decompiled (1 non-matching), some asm files split, etc

shawlucas opened the pull request:
I don't even recall everything I did, but basically its everything I have reserved on trello.

Also, it seems that preprocess.py runs faster when importing asm_processor as a module rather than calling it as subprocess. I'm not very good with python but I tried it (the same way that oot does it) and it was a bit faster (measure it with time command on linux). The makefile now explicitly calls python with "python3 -B" in order to prevent the creation of __pycache__ in the asm-processor folder.

Some of the early commits in this PR are from my previous PR, as I was working off that. Idk how that affects merging if at all (sorry im bad at git :( )

Rozelette commented on include/macros.h:
```diff
@@ -3,6 +3,9 @@
 
 #include "convert.h"
 
+#define true 1
```
These shouldn't go in `macros.h`. Instead, they should in `stdbool.h`. Copy what I have in #85

Rozelette commented on include/macros.h:
```diff
@@ -21,8 +24,8 @@
 #define SET_NEXT_GAMESTATE(curState, newInit, newStruct) \
-            (curState)->nextGameStateInit = (GameStateFunc)newInit; \
-            (curState)->nextGameStateSize = sizeof(newStruct);
+            (curState)->init = (GameStateFunc)newInit; \
```
I think you should keep the names for `nextGameStateInit` and `nextGameStateSize`. I think just calling them `init` and `size` is misleading, and it should be clear that they refer only to the _next_ game state

shawlucas replied:
But don't they point to whatever the init function and size of the current game state is? I mean it could be seen as next game state because if you set state->init = func itll use func as the current init function if im not mistaken. I could be wrong though. 

I think I was picturing it like this. If I was to look at the state in ram, lets say the current state is file context. so state->init would equal the file context init function, right? And then if you wanted to switch to the gameplay context, you could set the current init function to the gameplay init function. 

I could be completely wrong on this, its just what I thought and I can't check it in actual ram right now because I'm on my school computer lol

Rozelette replied:
@shawlucas Sorry for the late reply. The way that changing the gamestates work is that you set `nextGameStateInit` and `nextGameStateSize`, then set `isRunning` to 0. Then `GameState_Init` is called, which actually sets `nextGameStateInit` and `nextGameStateSize` both to 0. I think if you want know what the gamestate while the game is running, you'd look at the `main` function pointer.

Rozelette commented on src/code/game.c:
```diff
@@ -235,28 +239,28 @@ void Game_StateFini(GameState *ctxt) {
 
-GameStateFunc Game_GetNextStateInit(GameState *ctxt) {
-    return ctxt->nextGameStateInit;
+GameStateFunc GameState_GetInit(GameState* this) {
```
With the same logic as not renaming `nextGameStateInit`, I think this should be `GameState_GetNextStateInit`

m4xw replied:
+1

Rozelette reviewed (dismissed):
Thanks for another stack of files! LGTM except for a couple of names I disagree with

m4xw commented on src/code/game.c:
```diff
@@ -16,22 +16,22 @@ void Game_UpdateFramerateVariables(s32 divisor) {
 }
 
-void Game_SetFramerateDivisor(GameState *gameState, s32 divisor) {
+void Game_SetFramerateDivisor(GameState* this, s32 divisor) {
```
_this, same for func body

m4xw commented on src/code/game.c:
```diff
@@ -16,22 +16,22 @@ void Game_UpdateFramerateVariables(s32 divisor) {
 void GameState_SetFBFilter(Gfx** gfx, u32 arg1) {
-    Gfx* _gfx = *gfx;
+    Gfx* gfxP = *gfx;
 
```
Don't change or ptrGfx

shawlucas replied:
@Rozelette I kinda wanted more input on this because I really don't like the use of a prefixed _ in names of things. iirc, it also has issues with some compilers (i think we had this problem when dealing with object stuff in oot).

m4xw replied:
This only applies to names with 2 underscores or 1 underscore followed by a uppercase letter

m4xw replied:
(I dont like to use a shortform for the ptr since gfx is already a shortform.)

Kenix3 replied:
`gfxPtr` would be my ideal change here.

AngheloAlf replied:
`dlist` is used in this kind of stuff in other places, so you should consider it too.

m4xw commented on src/code/game.c:
```diff
@@ -55,160 +55,164 @@ void GameState_SetFBFilter(Gfx** gfx, u32 arg1) {
 }
 
-void Game_Nop80173534(GameState *ctxt) {
+void Game_Nop80173534(GameState* this) {
```
_this

m4xw commented on src/code/game.c:
```diff
@@ -55,160 +55,164 @@ void GameState_SetFBFilter(Gfx** gfx, u32 arg1) {
 }
 
-void GameState_Draw(GameState *ctxt, GraphicsContext *gCtxt) {
+void GameState_Draw(GameState* this, GraphicsContext *gfxCtx) {
```
_this

m4xw commented on src/code/game.c:
```diff
@@ -55,160 +55,164 @@ void GameState_SetFBFilter(Gfx** gfx, u32 arg1) {
 
-lblUnk:; // Label prevents reordering, if(1) around the above block don't seem to help unlike in OoT
-    func_800E9F78(gCtxt);
+label:
```
revert label name and comment, at least keep the reordering part

shawlucas replied:
That was an accident sorry

m4xw commented on src/code/game.c:
```diff
@@ -55,160 +55,164 @@ void GameState_SetFBFilter(Gfx** gfx, u32 arg1) {
-    gSPSegment(gCtxt->overlay.p++, 0, 0);
-    gSPSegment(gCtxt->overlay.p++, 0xF, gCtxt->framebuffer);
+void GameState_SetFrameBuffer(GraphicsContext *gfxCtx) {
+
```
Ignoring the name, remove the extra \n

m4xw commented on src/code/game.c:
```diff
@@ -55,160 +55,164 @@ void GameState_SetFBFilter(Gfx** gfx, u32 arg1) {
 
-void Game_UpdateInput(GameState *ctxt) {
-    Padmgr_GetInput(ctxt->input, 1);
+void Game_UpdateInput(GameState* this) {
```
_this

m4xw replied:
same for body

m4xw commented on src/code/game.c:
```diff
@@ -55,160 +55,164 @@ void GameState_SetFBFilter(Gfx** gfx, u32 arg1) {
-void Game_Update(GameState *ctxt) {
-    GraphicsContext* _gCtx;
-    _gCtx = ctxt->gfxCtx;
+void Game_Update(GameState* this) {
```
_this, same for body

m4xw commented on src/code/game.c:
```diff
@@ -55,160 +55,164 @@ void GameState_SetFBFilter(Gfx** gfx, u32 arg1) {
-void Game_IncrementFrameCount(GameState *ctxt) {
-    Game_Nop80173534(ctxt);
-    ctxt->frames++;
+void Game_IncrementFrameCount(GameState* this) {
```
_this, same for body

m4xw commented on src/code/game.c:
```diff
@@ -55,160 +55,164 @@ void GameState_SetFBFilter(Gfx** gfx, u32 arg1) {
+void GameState_Realloc(GameState* this, size_t size) {
+    GameAlloc* alloc;
+    void* gameArena;
+    size_t systemMaxFree;
```
Adding to roz' comment, please keep size_t :)

m4xw commented on src/code/game.c:
```diff
@@ -55,160 +55,164 @@ void GameState_SetFBFilter(Gfx** gfx, u32 arg1) {
-    gCtxt->viConfigYScale = gViConfigYScale;
-    ctxt->nextGameStateInit = NULL;
-    ctxt->nextGameStateSize = 0U;
+void GameState_Init(GameState* this, GameStateFunc init, GraphicsContext* gfxCtx) {
```
_this etc

m4xw commented on src/code/game.c:
```diff
@@ -235,28 +239,28 @@ void Game_StateFini(GameState *ctxt) {
-s32 func_80173B48(GameState *ctxt) {
-    s32 result;
-    result = OS_CYCLES_TO_NSEC(ctxt->framerateDivisor * sIrqMgrRetraceTime) - OS_CYCLES_TO_NSEC(D_801FBAF0);
+s32 func_80173B48(GameState* this) {
```
_this, same for body

m4xw commented on src/code/game.c:
```diff
@@ -235,28 +239,28 @@ void Game_StateFini(GameState *ctxt) {
 
-s32 Game_GetHeapFreeSize(GameState *ctxt) {
-    return THA_GetSize(&ctxt->heap);
+s32 GameState_GetArenaSize(GameState* this) {
```
_this, same for body

m4xw commented on src/code/game.c:
```diff
@@ -235,28 +239,28 @@ void Game_StateFini(GameState *ctxt) {
 
-u32 Game_GetShouldContinue(GameState *ctxt) {
-    return ctxt->running;
+u32 GameState_IsRunning(GameState* this) {
```
_this, same for body, name should be considered again, i will leave that open

m4xw commented on src/code/game.c:
```diff
@@ -235,28 +239,28 @@ void Game_StateFini(GameState *ctxt) {
 
-u32 Game_GetNextStateSize(GameState *ctxt) {
-    return ctxt->nextGameStateSize;
+size_t Game_GetNextStateSize(GameState* this) {
```
_this, same for body
size_t is good

m4xw reviewed (changes requested):
Thats my nits so far

Rozelette commented:
The way merging works for maintainers is I squash and merge, meaning all the commits become one that's then merged with master. Maintainers can rebase and merge, but I don't generally like that because I don't see the need to keep track of all the intermediate commits before the PR is done.

Also, to clarify something earlier, because I squash and merge, you don't strictly have to rebase off of master. You can just merge master to deal with conflicts. So if that's more comfortable to you, feel free to do so.

shawlucas commented on src/boot_O2_g3_trapuv/fault_drawer.c:
```diff
@@ -1,5 +1,39 @@
+#include <vt.h>
+
+// rodata
+const u32 sFaultDrawerFont[] = {
```
Okay ill take a look

Kenix3 replied:
If you're having trouble with this, I'd suggest just extracting as a blob. It's very possible it's not in a texture format.

shawlucas commented:
I'll update this once tharo's new build system is finished because it'll be a lot easier to deal with conflicts.

AngheloAlf commented:
Just to let you know, the new build system has been merged, so you can now continue working on this :D

shawlucas commented:
Ok so i fixed all the merge conflicts. The only thing that isn't working right now is in z_collision_check, there are two fake bss symbols. Keeping these in makes it not match. Removing them makes it match, but then it creates an undefined reference error in z_message_nes. So I added the symbol to undefined_syms.txt, idk what else to do

EllipticEllipsis commented:
You should be able to edit `variables.txt` with the new data type that swallows up the fake symbols, and if you disassemble again, it should put the correct symbol + offset in place in the asm instead. That's meant to be how NBS works.

shawlucas commented:
@EllipticEllipsis i have this in variables.txt: ``    0x801EF66D:("D_801EF66D","UNK_TYPE1","",0x1),`` but for some reason i still get an undefined reference error: 
```
mips-linux-gnu-ld: build/src/code/z_message_nes.o: in function `func_8015B198':
(.text+0x453c): undefined reference to `D_801EF66D'
```

EllipticEllipsis commented:
Is that symbol the fake one?

shawlucas commented:
> Is that symbol the fake one?

yes

shawlucas commented:
and i get undefined reference unless i have D_801EF66D = 0x801EF66D; in undefined_syms.txt

shawlucas commented:
The symbol shows up in z_collision_check.bss.s but its "used" in z_message_nes. trying to add the symbol in z_collision_check puts it before any of the other bss symbols, even though it should be at the end

shawlucas commented:
@EllipticEllipsis this is what happens in the map:
```
 .bss           0x00000000801ede00     0x1880 build/src/code/z_collision_check.o
                0x00000000801ede00                D_801EF66D
                0x00000000801ede08                D_801EDE00
                0x00000000801ede18                D_801EDE10
                0x00000000801ede28                D_801EDE20
```
when I put the symbol at the end of all the other bss in z_collision_check.c:

```
...
TriNorm D_801EF5C8;
TriNorm D_801EF600;
TriNorm D_801EF638;
UNK_TYPE1 D_801EF66D;
```

EllipticEllipsis commented:
Ugh, what a pain. Can't even use the hack I found to put it in its own file since the alignment's wrong... unless... Is there room above it to align it correctly from a new file boundary?

shawlucas commented:
the previous bss symbol is a TriNorm, and it looks like D_801EF66D should be part of D_801EF638. The trinorm is 0x34 in size, and padding makes is 0x38 (shown in all the previous trinorm symbols). However, D_801EF638 in the .bss file shows as .space 0x35, and then D_801EF66D being .space 0x3.

EllipticEllipsis commented:
Ugh, that's really bad. I was hoping there'd be more room than that. Hmm... What happens if you un-import the bss?

shawlucas commented:
It matches if I include z_collision_check.bss.o in the spec and then extern all the bss in z_collision_check file...but then that means that we are not migrating collision check bss.

EllipticEllipsis commented:
We'll have to live with that for now, I think. Last suggestion: all the bss in its own C file, included using the spec?

shawlucas commented:
nope lol

engineer124 commented on src/code/z_skelanime.c:
```diff
@@ -1,659 +1,663 @@
-            gSPMatrix(polyTemp, SysMatrix_GetStateAsRSPMatrix(mtx), G_MTX_LOAD);
-            gSPDisplayList(polyTemp + 1, dList[1]);
-            gfxCtx->polyOpa.p = polyTemp + 2;
+            gSPMatrix(polyTemp, Matrix_ToMtx(mtx), G_MTX_LOAD);
+            gSPDisplayList(polyTemp + 1, newDList);
+            POLY_OPA_DISP = polyTemp + 2;
```
```suggestion
            gSPMatrix(&polyTemp[0], Matrix_ToMtx(mtx), G_MTX_LOAD);
            gSPDisplayList(&polyTemp[1], newDList);
            POLY_OPA_DISP = &polyTemp[2];
```

engineer124 commented on src/code/z_skelanime.c:
```diff
@@ -1,659 +1,663 @@
-            Gfx* polyTemp = gfxCtx->polyOpa.p;
+            Gfx* polyTemp = POLY_OPA_DISP;
 
             gSPMatrix(polyTemp, Matrix_NewMtx(globalCtx->state.gfxCtx), G_MTX_LOAD);
             gSPDisplayList(polyTemp + 1, dList);
```
```suggestion
            gSPMatrix(&polyTemp[0], Matrix_NewMtx(globalCtx->state.gfxCtx), G_MTX_LOAD);
            gSPDisplayList(&polyTemp[1], dList);
```

engineer124 commented on src/code/z_skelanime.c:
```diff
@@ -1,659 +1,663 @@
             gSPMatrix(polyTemp, Matrix_NewMtx(globalCtx->state.gfxCtx), G_MTX_LOAD);
             gSPDisplayList(polyTemp + 1, dList);
-            gfxCtx->polyOpa.p = polyTemp + 2;
+            POLY_OPA_DISP = polyTemp + 2;
```
```suggestion
            POLY_OPA_DISP = &polyTemp[2];
```

engineer124 commented on src/code/z_skelanime.c:
```diff
@@ -1,659 +1,663 @@
-            Gfx* polyTemp = gfxCtx->polyOpa.p;
+            Gfx* polyTemp = POLY_OPA_DISP;
 
             gSPMatrix(polyTemp, Matrix_NewMtx(globalCtx->state.gfxCtx), G_MTX_LOAD);
             gSPDisplayList(polyTemp + 1, dList);
```
```suggestion
            gSPMatrix(&polyTemp[0], Matrix_NewMtx(globalCtx->state.gfxCtx), G_MTX_LOAD);
            gSPDisplayList(&polyTemp[1], dList);
```

engineer124 commented on src/code/z_skelanime.c:
```diff
@@ -1,659 +1,663 @@
             gSPMatrix(polyTemp, Matrix_NewMtx(globalCtx->state.gfxCtx), G_MTX_LOAD);
             gSPDisplayList(polyTemp + 1, dList);
-            gfxCtx->polyOpa.p = polyTemp + 2;
+            POLY_OPA_DISP = polyTemp + 2;
```
```suggestion
            POLY_OPA_DISP = &polyTemp[2];
```

engineer124 commented on src/code/z_skelanime.c:
```diff
@@ -1,659 +1,663 @@
-            gSPMatrix(polyTemp, SysMatrix_GetStateAsRSPMatrix(mtx), G_MTX_LOAD);
-            gSPDisplayList(polyTemp + 1, dList[1]);
-            gfxCtx->polyOpa.p = polyTemp + 2;
+            gSPMatrix(polyTemp, Matrix_ToMtx(mtx), G_MTX_LOAD);
+            gSPDisplayList(polyTemp + 1, newDList);
+            POLY_OPA_DISP = polyTemp + 2;
```
```suggestion
            gSPMatrix(&polyTemp[0], Matrix_ToMtx(mtx), G_MTX_LOAD);
            gSPDisplayList(&polyTemp[1], newDList);
            POLY_OPA_DISP = &polyTemp[2];
```

engineer124 commented on src/code/z_skelanime.c:
```diff
@@ -1,659 +1,663 @@
-            gSPMatrix(polyTemp, SysMatrix_GetStateAsRSPMatrix(*mtx), G_MTX_LOAD);
-            gSPDisplayList(polyTemp + 1, dList[1]);
-            gfxCtx->polyOpa.p = polyTemp + 2;
+            gSPMatrix(polyTemp, Matrix_ToMtx(*mtx), G_MTX_LOAD);
+            gSPDisplayList(polyTemp + 1, newDList);
+            POLY_OPA_DISP = polyTemp + 2;
```
```suggestion
            gSPMatrix(&polyTemp[0], Matrix_ToMtx(*mtx), G_MTX_LOAD);
            gSPDisplayList(&polyTemp[1], newDList);
            POLY_OPA_DISP = &polyTemp[2];
```

engineer124 commented on src/code/z_skelanime.c:
```diff
@@ -1,659 +1,663 @@
-            gSPMatrix(polyTemp, SysMatrix_GetStateAsRSPMatrix(mtx), G_MTX_LOAD);
-            gSPDisplayList(polyTemp + 1, dList[1]);
-            gfxCtx->polyOpa.p = polyTemp + 2;
+            gSPMatrix(polyTemp, Matrix_ToMtx(mtx), G_MTX_LOAD);
+            gSPDisplayList(polyTemp + 1, newDList);
+            POLY_OPA_DISP = polyTemp + 2;
```
```suggestion
            gSPMatrix(&polyTemp[0], Matrix_ToMtx(mtx), G_MTX_LOAD);
            gSPDisplayList(&polyTemp[1], newDList);
            POLY_OPA_DISP = &polyTemp[2];
```

engineer124 commented on src/code/z_skelanime.c:
```diff
@@ -717,205 +722,288 @@ Gfx* SkelAnime_Draw2(GlobalContext* globalCtx, void** skeleton, Vec3s* limbDrawT
+    if ((overrideLimbDraw == NULL) || !overrideLimbDraw(globalCtx, limbIndex, &newDList, &pos, &rot, arg, &gfx)) {
+        Matrix_JointPosition(&pos, &rot);
+        if (newDList != NULL) {
+            gSPMatrix(gfx, Matrix_ToMtx(*mtx), G_MTX_LOAD);
+            gSPDisplayList(gfx + 1, newDList);
             gfx = gfx + 2;
```
```suggestion
            gSPMatrix(&gfx[0], Matrix_ToMtx(*mtx), G_MTX_LOAD);
            gSPDisplayList(&gfx[1], newDList);
            gfx = &gfx[2];
```
I think I'll stop commenting on these from here, there's a lot of these in the file to change

engineer124 commented on src/code/z_skelanime.c:
```diff
@@ -929,807 +1017,997 @@ AnimationEntry* SkelAnime_NextEntry(AnimationContext* animationCtx, AnimationTyp
+ * animation frame
+ */
+s32 SkelAnime_MorphTaper(SkelAnime* skelAnime) {
+    s16 prevPhase = (s16)(skelAnime->morphWeight * 16384.0f);
```
```suggestion
    s16 prevPhase = (s16)(skelAnime->morphWeight * (f32)0x4000);
```
What do you think about this? weight seems like it should be in dec, but it's quite a nice hex number

shawlucas replied:
Found out that i dont need the s16 cast or the f32 cast and can just use the hex.

engineer124 commented on src/code/z_skelanime.c:
```diff
@@ -929,807 +1017,997 @@ AnimationEntry* SkelAnime_NextEntry(AnimationContext* animationCtx, AnimationTyp
-                               skelAnime->limbDrawTbl);
-        func_80136F04(skelAnime);
+    if (skelAnime->curFrame == skelAnime->endFrame) {
+        SkelAnime_GetFrameData(skelAnime->animation, (s32)skelAnime->curFrame, skelAnime->limbCount,
```
```suggestion
        SkelAnime_GetFrameData(skelAnime->animation, skelAnime->curFrame, skelAnime->limbCount,
```
Is the s32 cast needed?

engineer124 commented on src/libultra/io/contpfs.c:
```diff
@@ -0,0 +1,308 @@
+#include <ultra64.h>
+#include <global.h>
+
```
```suggestion
#include "ultra64.h"
#include "global.h"

```

engineer124 commented on src/libultra/io/epilinkhandle.c:
```diff
@@ -0,0 +1,13 @@
+#include <ultra64.h>
+#include <global.h>
```
```suggestion
#include "ultra64.h"
#include "global.h"
```

engineer124 commented on src/libultra/io/motor.c:
```diff
@@ -0,0 +1,120 @@
+#include <ultra64.h>
+#include <global.h>
```
```suggestion
#include "ultra64.h"
#include "global.h"
```

engineer124 commented on src/libultra/libc/xlitob.c:
```diff
@@ -0,0 +1,4 @@
+#include <ultra64.h>
+#include <global.h>
```
```suggestion
#include "ultra64.h"
#include "global.h"
```

engineer124 commented on src/libultra/voice/voicegetreaddata.c:
```diff
@@ -0,0 +1,4 @@
+#include <ultra64.h>
+#include <global.h>
```
```suggestion
#include "ultra64.h"
#include "global.h"
```

engineer124 reviewed (commented):
Great work with this! It'll be really nice to have the documented animation in the repo. Just mostly styling comments

AngheloAlf commented:
More conflicts here :c

shawlucas commented on include/variables.h:
```diff
@@ -3368,13 +3367,10 @@ extern s32* gNMIBuffer;
-extern s32 D_801F5AB0;
-extern s32 D_801F5AB4;
+extern QuakeRequest sQuakeRequest[4];
+extern Quake2Context sQuake2Context;
```
Nope

EllipticEllipsis commented:
New idea: can you use the SM64 `prevent_bss_reordering` header? It's worked in other files. Add it to `includes/` and include it right at the top of the file with the problematic bss.

EllipticEllipsis commented on spec:
```diff
@@ -478,6 +477,7 @@ beginseg
     include "build/src/code/z_collision_btltbls.o"
     pad_text
     include "build/src/code/z_collision_check.o"
+  //  include "build/data/code/z_collision_check.bss.o"
```
You can get rid of this now.

EllipticEllipsis commented on src/boot_O2_g3/fault.c:
```diff
@@ -1,7 +1,35 @@
 #include "vt.h"
 
-#pragma GLOBAL_ASM("asm/non_matchings/boot/fault/D_800982B0.s")
+// TODO move out
+#define OS_CLOCK_RATE 62500000LL
```
This is in a header already, I think?

EllipticEllipsis commented on src/code/z_en_a_keep.c:
```diff
@@ -1,27 +1,27 @@
-    ActorEnAObj* s0 = (ActorEnAObj*)this;
+void EnAObj_Init(EnAObj* thisx, GlobalContext* globalCtx) {
+    EnAObj* s0 = (EnAObj*)thisx;
     s0->base.textId = ((s0->base.params >> 8) & 0xFF) | 0x300;
     s0->base.params = (s0->base.params & 0xFF) - 9;
     Actor_ProcessInitChain((Actor*)s0, &enAObjInitVar);
     ActorShape_Init(&s0->base.shape, 0, (ActorShadowFunc)func_800B3FC0, 12);
     Collider_InitAndSetCylinder(globalCtx, &s0->collision, (Actor*)s0, &enAObjCylinderInit);
     Collider_UpdateCylinder((Actor*)s0, &s0->collision);
     s0->base.colChkInfo.mass = 255;
-    s0->update = (ActorFunc)EnAObj_Update1;
+    s0->actionFunc = (EnAObjActionFunc)EnAObj_Update1;
```
`s0 -> this` throughout.

EllipticEllipsis commented on src/code/z_en_a_keep.c:
```diff
@@ -1,27 +1,27 @@
 }
 
-void EnAObj_Destroy(ActorEnAObj* this, GlobalContext* globalCtx) {
+void EnAObj_Destroy(EnAObj* this, GlobalContext* globalCtx) {
     ColliderCylinder* a2 = &this->collision;
     Collider_DestroyCylinder(globalCtx, a2);
 }
```
```suggestion
void EnAObj_Destroy(Actor* thisx, GlobalContext* globalCtx) {
    EnAObj* this = (EnAObj*)thisx
    Collider_DestroyCylinder(globalCtx, &this->collision);
}
```

EllipticEllipsis commented on src/code/z_en_a_keep.c:
```diff
@@ -1,27 +1,27 @@
+void EnAObj_Update1(EnAObj* this, GlobalContext* globalCtx) {
     s16 v0;
     s32 v1;
     if (func_800B84D0((Actor*)this, globalCtx) != 0) {
```
```suggestion
    if (func_800B84D0(&this->actor, globalCtx) != 0) {
```
Maybe a boolean, too?

EllipticEllipsis commented on src/code/z_en_a_keep.c:
```diff
@@ -1,27 +1,27 @@
     s32 v1;
     if (func_800B84D0((Actor*)this, globalCtx) != 0) {
-        this->update = (ActorFunc)EnAObj_Update2;
+        this->actionFunc = (EnAObjActionFunc)EnAObj_Update2;
```
```suggestion
        this->actionFunc = EnAObj_Update2;
```

EllipticEllipsis commented on src/code/z_en_a_keep.c:
```diff
@@ -31,18 +31,18 @@ void EnAObj_Update1(ActorEnAObj* this, GlobalContext* globalCtx) {
 
-void EnAObj_Update2(ActorEnAObj* this, GlobalContext* globalCtx) {
+void EnAObj_Update2(EnAObj* this, GlobalContext* globalCtx) {
     if (func_800B867C((Actor*)this, globalCtx) != 0) {
```
```suggestion
    if (func_800B867C(&this->actor, globalCtx) != 0) {
```

EllipticEllipsis commented on src/code/z_en_a_keep.c:
```diff
@@ -31,18 +31,18 @@ void EnAObj_Update1(ActorEnAObj* this, GlobalContext* globalCtx) {
+void EnAObj_Update2(EnAObj* this, GlobalContext* globalCtx) {
     if (func_800B867C((Actor*)this, globalCtx) != 0) {
-        this->update = (ActorFunc)EnAObj_Update1;
+        this->actionFunc = (EnAObjActionFunc)EnAObj_Update1;
```
```suggestion
        this->actionFunc = EnAObj_Update1;
```

EllipticEllipsis commented on src/code/z_en_a_keep.c:
```diff
@@ -31,18 +31,18 @@ void EnAObj_Update1(ActorEnAObj* this, GlobalContext* globalCtx) {
 
-void EnAObj_Update(ActorEnAObj* this, GlobalContext* globalCtx) {
-    (this->update)((Actor*)this, (GlobalContext*)globalCtx);
+void EnAObj_Update(EnAObj* this, GlobalContext* globalCtx) {
+    (this->actionFunc)((struct EnAObj*)this, (GlobalContext*)globalCtx);
     Actor_SetHeight((Actor*)this, 45.0f);
     CollisionCheck_SetOC(globalCtx, &globalCtx->colChkCtx, (Collider*)&this->collision);
 }
 
```
Usual recasting, and actionFunc call can surely be simplified.

EllipticEllipsis commented on src/code/z_en_a_keep.c:
```diff
@@ -31,18 +31,18 @@ void EnAObj_Update1(ActorEnAObj* this, GlobalContext* globalCtx) {
 }
 
-void EnAObj_Draw(ActorEnAObj* this, GlobalContext* globalCtx) {
+void EnAObj_Draw(EnAObj* this, GlobalContext* globalCtx) {
     func_800BDFC0(globalCtx, enAObjDisplayLists[this->base.params]);
 }
```
`thisx`, and recast if room.

shawlucas replied:
Regalloc gets messed up if i recast here, so just using the thisx

EllipticEllipsis commented on src/code/z_fcurve_data_skelanime.c:
```diff
@@ -1,15 +1,181 @@
+                    if (j == 0) {
+                        *transforms = transformValue * 1024.0f;
+                    } else if (j == 1) {
+                        *transforms = transformValue * D_801DD7C0; /* (32768.0f / 180.0f); */
```
Should this be migrated too?

EllipticEllipsis commented on src/code/z_skelanime.c:
```diff
@@ -1,680 +1,684 @@
+    if ((overrideLimbDraw == NULL) || !overrideLimbDraw(globalCtx, limbIndex, &newDList, &pos, &rot, arg)) {
+        Matrix_JointPosition(&pos, &rot);
+        Matrix_StatePush();
+        unkDraw(globalCtx, limbIndex, arg);
```
```suggestion
        //! @bug Does not check unkDraw is not NULL before calling it.
        unkDraw(globalCtx, limbIndex, arg);
```
And elsewhere too in this file.

EllipticEllipsis commented on src/libultra/io/devmgr.c:
```diff
@@ -1,3 +1,109 @@
 #include "global.h"
 
+#ifdef NON_MATCHING
+// rodata issues
```
What specifically?

EllipticEllipsis commented on src/libultra/io/motor.c:
```diff
@@ -0,0 +1,120 @@
+#include "ultra64.h"
+#include "global.h"
+
```
Is this the correct name for this file? It's currently called `rumblepak.c`.

shawlucas replied:
libultra used motor in place of pak/rumble/mempak. the file deals with both the rumblepak and memory pak so idk maybe name the file pak.c ?

EllipticEllipsis commented on undefined_syms.txt:
```diff
@@ -4184,3 +4184,5 @@ D_0E0002F8 = 0x0E0002F8;
 
 D_0F000000 = 0x0F000000;
+
+// fake bss
```
Is this still needed?

EllipticEllipsis commented on src/overlays/actors/ovl_En_Toto/z_en_toto.c:
```diff
@@ -205,8 +205,8 @@ void EnToto_Init(Actor* thisx, GlobalContext* globalCtx) {
-    SkelAnime_InitSV(globalCtx, &this->skelAnime, &D_0600A978, globalCtx->sceneNum == 0x12 ? &D_06003AA8 : &D_0600C880,
-                     this->limbDrawTbl, this->transitionDrawTbl, 18);
+    SkelAnime_InitFlex(globalCtx, &this->skelAnime, &D_0600A978,
+                       globalCtx->sceneNum == 0x12 ? &D_06003AA8 : &D_0600C880, this->jointTable, this->morphTable, 18);
```
```suggestion
                       ( (globalCtx->sceneNum == 0x12) ? &D_06003AA8 : &D_0600C880 ), this->jointTable, this->morphTable, 18);
```
This is hard enough to read in the first place...

EllipticEllipsis reviewed (commented):
I'm looking forward to getting all this SkelAnime documentation in. I just have a few suggestions for cleanup, and a few other things.

One general concern: you haven't been very consistent about `void* thisx` vs `void* arg` in the actor LimbDraws. I'd rather it be one or the other. Actually, I'm not entirely convinced that we even need `void*` in these functions in MM yet, although in principle they don't need to take an `Actor*`.

Also, I don't think these matrix function names are final anyway, right?

shawlucas commented:
well i already changed all the actors to use void* for limb draw and postlimbdraw functions so I wouldn't get a warning lol

AngheloAlf commented:
Have we ever seen any limbdraw function which doesn't take an actor for that parameter? I think we haven't, in which case I think leaving it as `Actor*` should be the correct way imo

AngheloAlf commented on src/boot_O2_g3/fault.c:
```diff
@@ -712,9 +737,10 @@ void Fault_ProcessClients(void) {
-            FaultDrawer_Printf("\x1A\x38"
-                               "CallBack (%d) %08x %08x %08x\n"
-                               "\x1A\x37",
+            FaultDrawer_Printf("\x1a"
+                               "8CallBack (%d) %08x %08x %08x\n"
+                               "\x1a"
+                               "7",
```
I think what was before was better here since those are escape characters

AngheloAlf commented on src/boot_O2_g3/fault.c:
```diff
@@ -869,7 +897,7 @@ void Fault_Start(void) {
     gFaultStruct.faultHandlerEnabled = 1;
     osCreateMesgQueue(&sFaultContext->queue, sFaultContext->msg, 1);
-    StackCheck_Init(&sFaultThreadInfo, sFaultStack, sFaultStack + sizeof(sFaultStack), 0, 0x100, "fault");
+    StackCheck_Init(&sFaultThreadInfo, sFaultStack, sFaultStack + sizeof(sFaultStack), 0, 0x100, faultThreadName);
```
Any special reason for this change?

AngheloAlf commented on src/code/code_80122660.c:
```diff
@@ -1,5 +1,5 @@
 
-#pragma GLOBAL_ASM("asm/non_matchings/code/code_80122660/func_80122660.s")
+#pragma GLOBAL_ASM("asm/non_matchings/code/z_pause/func_80122660.s")
 
-#pragma GLOBAL_ASM("asm/non_matchings/code/code_80122660/func_80122670.s")
+#pragma GLOBAL_ASM("asm/non_matchings/code/z_pause/func_80122670.s")
```
I see you made a z_pause file, so you should delete this file to avoid duplication.

AngheloAlf commented on src/code/game.c:
```diff
@@ -54,28 +54,26 @@ void GameState_SetFBFilter(Gfx** gfx, u32 arg1) {
 }
 
-void Game_Nop80173534(GameState* gamestate) {
+void Game_Nop80173534(GameState* gameState) {
     ;
 }
```
Is the semicolon needed?
```suggestion
void Game_Nop80173534(GameState* gameState) {
}
```

AngheloAlf commented on src/code/z_eff_footmark.c:
```diff
@@ -106,8 +106,8 @@ void EffFootmark_Draw(GlobalContext* globalCtx) {
-            SysMatrix_SetCurrentState(&footmark->displayMatrix);
-            Matrix_Scale(footmark->size * 0.00390625f * 0.7f, 1, footmark->size * 0.00390625f, 1);
+            Matrix_SetCurrentState(&footmark->displayMatrix);
+            Matrix_Scale(footmark->size * 0.00390625f * 0.7f, 1, footmark->size * 0.00390625f, MTXMODE_APPLY); 
```
```suggestion
            Matrix_Scale(footmark->size * (1.0f / 0x100) * 0.7f, 1, footmark->size * (1.0f / 0x100), MTXMODE_APPLY); 
```

AngheloAlf commented on src/code/z_effect_soft_sprite.c:
```diff
@@ -69,10 +69,18 @@ void EffectSS_ResetEntry(EffectSs* particle) {
-    particle->velocity.x = particle->velocity.y = particle->velocity.z = 0;
-    particle->vec.x = particle->vec.y = particle->vec.z = 0;
-    particle->pos.x = particle->pos.y = particle->pos.z = 0;
+    particle->accel.z = 0;
+    particle->accel.y = 0;
+    particle->accel.x = 0;
+    particle->velocity.z = 0;
+    particle->velocity.y = 0;
+    particle->velocity.x = 0;
+    particle->vec.z = 0;
+    particle->vec.y = 0;
+    particle->vec.x = 0;
+    particle->pos.z = 0;
+    particle->pos.y = 0;
+    particle->pos.x = 0;
```
What was before wasn't bad imo

AngheloAlf commented on src/code/z_effect_soft_sprite_old_init.c:
```diff
@@ -402,8 +402,9 @@ void EffectSsGSpk_SpawnSmall(GlobalContext* globalCtx, Actor* actor, Vec3f* pos,
 }
 
 // EffectSsDFire Spawn Functions
+
 void EffectSsDFire_Spawn(GlobalContext* globalCtx, Vec3f* pos, Vec3f* velocity, Vec3f* accel, s16 scale, s16 scaleStep,
```
```suggestion
// EffectSsDFire Spawn Functions
void EffectSsDFire_Spawn(GlobalContext* globalCtx, Vec3f* pos, Vec3f* velocity, Vec3f* accel, s16 scale, s16 scaleStep,
```

AngheloAlf commented on src/libultra/libc/xlitob.c:
```diff
@@ -0,0 +1,4 @@
+#include "ultra64.h"
+#include "global.h"
+
+#pragma GLOBAL_ASM("asm/non_matchings/boot/xlitob/_Litob.s")
```
This file is a duplicated of `src/libultra/rmon/xlitob.c`

AngheloAlf commented on Makefile:
```diff
@@ -255,9 +256,12 @@ build/src/overlays/%.o: src/overlays/%.c
 
+build/src/%.o: %.c
+	$(CC) -c $(CFLAGS) $(MIPS_VERSION) $(OPTFLAGS) -o $@ $<
+
```
Whoops, this is outdated :sweat_smile:

shawlucas commented:
I can work on this tomorrow, haven't had time recently

shawlucas commented:
So I started college and really haven't had time to work on this. Hopefully tonight. I feel really bad about the delay I have just been having so much college work to do

AngheloAlf commented:
We discussed the `Actor*` vs `void*` thing in discord. After checking every skelanime draw function, all of them pass either an `Actor*` or `NULL`, so we decided to use `Actor*` instead of `void*`.
To avoid you to change all of them back, I changed them and I opened a PR to your branch (https://github.com/shawlucas/mm/pull/1)

I think we can merge this PR as soon as you merge that PR.

shawlucas commented:
merged pr
